[Concurrency] Add `async` to the Swift type system.

DougGregor opened the pull request:
Add `async` to the type system. `async` can be written as part of a
function type or function declaration, following the parameter list, e.g.,

  func doSomeWork() async { ... }

`async` functions are distinct from non-`async` functions and there
are no conversions amongst them. At present, `async` functions do not
*do* anything, but this commit fully supports them as a distinct kind
of function throughout:

* Parsing of `async`
* AST representation of `async` in declarations and types
* Syntactic type representation of `async`
* (De-/re-)mangling of function types involving 'async'
* Runtime type representation and reconstruction of function types
involving `async`.
* Dynamic casting restrictions for `async` function types
* (De-)serialization of `async` function types
* Disabling overriding, witness matching, and conversions with
differing `async`

DougGregor committed:
[Concurrency] Add `async` to the Swift type system.

Add `async` to the type system. `async` can be written as part of a
function type or function declaration, following the parameter list, e.g.,

  func doSomeWork() async { ... }

`async` functions are distinct from non-`async` functions and there
are no conversions amongst them. At present, `async` functions do not
*do* anything, but this commit fully supports them as a distinct kind
of function throughout:

* Parsing of `async`
* AST representation of `async` in declarations and types
* Syntactic type representation of `async`
* (De-/re-)mangling of function types involving 'async'
* Runtime type representation and reconstruction of function types
involving `async`.
* Dynamic casting restrictions for `async` function types
* (De-)serialization of `async` function types
* Disabling overriding, witness matching, and conversions with
differing `async`

rjmccall commented on include/swift/AST/Decl.h:
```diff
@@ -5899,9 +5907,15 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   SourceLoc getThrowsLoc() const { return ThrowsLoc; }
 
+  /// Returns true if the function is async.
+  bool hasAsync() const { return Bits.AbstractFunctionDecl.Async; }
```
Some of the features we're considering distinguish between functions that behave like `async` functions for their callers and those that expose an `async` context within the implementation.  (This is useful for e.g. UI event handlers.) This should probably be clear about which it means.

rjmccall commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -381,6 +381,10 @@ def enable_experimental_static_assert :
 
+def enable_experimental_async :
+  Flag<["-"], "enable-experimental-async">,
+  HelpText<"Enable experimental async concurrency">;
```
Can you name this `-enable-experimental-concurrency`, since the feature is significantly broader?

rjmccall commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -7411,14 +7414,23 @@ Parser::parseDeclInit(ParseDeclOptions Flags, DeclAttributes &Attributes) {
     Attributes.add(new (Context) RethrowsAttr(throwsLoc));
   }
 
+  // Initializers cannot be 'async'.
```
Maybe leave a TODO saying that this should be conditionally lifted.

rjmccall reviewed (commented):
Mostly looks great, thanks!

sagesse-cn commented:
I think syntax format
```swift
async func doSomeWork () {}
// or
@async func doSomeWork () {}
```
Is a better choice

prtmshk commented:
I agree with @sagesse-cn. `async func` reads more naturally than `func functionName() async`. Consider a throwing async function. Where would the `async` keyword go, before or after the `throws` keyword? `async func functionName() throws -> ReturnType` reads more naturally than `func functionName() async throws -> ReturnType` or `func functionName() throws async -> ReturnType` or even `func functionName() throws -> async ReturnType`

minuscorp commented:
It is in the code about how the semantics are:
`"'async' must precede %select{'throws'|'rethrows'}0", (bool))`

kylebrowning commented:
Would be nice to define a thread to run the async func on. 

func myFunction async(.main) {

ScottRobbins commented:
I assume there will be a SE post in the future to bikeshed syntax.

CodaFi commented on lib/Demangling/Remangler.cpp:
```diff
@@ -2314,6 +2314,10 @@ void Remangler::mangleFullObjCResilientClassStub(Node *node) {
   Buffer << "Mt";
 }
 
+void Remangler::mangleAsyncAnnotation(Node *node) {
```
When you have the opportunity, please update the [mangling reference](https://github.com/apple/swift/blob/master/docs/ABI/Mangling.rst)

DougGregor replied:
I did... https://github.com/apple/swift/pull/33147/files#diff-3d09d8bf538c5efb52870cdf565b219c

CodaFi commented on lib/Sema/TypeCheckProtocol.h:
```diff
@@ -234,6 +234,9 @@ enum class MatchKind : uint8_t {
   MissingRequirement,
 
+  /// The witness is async, but the requirement does not.
+  AsyncConflict,
```
Nit: `is not async`

juliand665 commented:
While I agree that `async func name() -> Value` reads more naturally, `async` to me feels analogous to `throws` in that it specifies how the function returns, and should thus directly precede the `->`: `func name() async -> Value`. You could read this similarly too, "a `func`tion named `name` taking no parameters (`()`), `async`hronously returning (`->`) a `Value`".

The point is it's the return (`->`) that's asynchronous, not the function itself—as opposed to, say, `DispatchQueue.main.async`, which functionally doesn't do anything until a later time.

theblixguy commented:
I don’t think this PR is an appropriate place to discuss syntax changes. Please move the discussion to the Swift forums. The most recent thread around `async` is [here](https://forums.swift.org/t/rethinking-async/37984).

DougGregor committed:
Fix tests for slightly altered Fix-Its for throws/rethrows parsing.

DougGregor committed:
[Concurrency] Rename `-enable-experimental-async`.

Use the more general name `-enable-experimental-concurrency` instead,
since async is one part of the model.

DougGregor committed:
[Concurrency] Improve comments based on review feedback.

lattner commented:
> I think syntax format
> 
> ```swift
> async func doSomeWork () {}
> // or
> @async func doSomeWork () {}
> ```
> 
> Is a better choice

Keep in mind that async needs to work with function types, which are postfix.  This is the same issue as the existing `throws` effect, so it should be modeled the same way.

DougGregor committed:
[Concurrency] Add Syntax support for 'async'.

Add 'async' to the syntax tree for types, function declarations, and
the "arrow" expression.

DougGregor committed:
[Concurrency] Finish and add tests for ban on `async` with `@objc`.

DougGregor commented:
Editorial note: I've deleted all of the comments bike-shedding syntax. Those belong on the forums, in a pitch thread about concurrency designs, not on a pull request for the implementation.

DougGregor commented:
@swift-ci please smoke test

https://github.com/apple/swift-syntax/pull/230

DougGregor commented:
@swift-ci please smoke test

apple/swift-syntax#230

DougGregor commented:
https://github.com/apple/swift-syntax/pull/230

@swift-ci smoke test

DougGregor commented:
SwiftSyntax part of this over at https://github.com/apple/swift-syntax/pull/230

rintaro commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -783,40 +784,18 @@ Parser::parseFunctionSignature(Identifier SimpleName,
-    }
+  // Check for the 'async' keyword.
+  if (Context.LangOpts.EnableExperimentalConcurrency &&
+      Tok.isContextualKeyword("async")) {
+    asyncLoc = consumeToken();
+  }
 
-    if (!throwsLoc.isValid())
-      return None;
+  // Check for the 'async' and 'throws' keywords.
+  rethrows = false;
+  parseAsyncThrows(SourceLoc(), asyncLoc, throwsLoc, &rethrows);
```
788-792 is a leftover? If it's really needed, could you clarify why it can't be parsed in `parseAsyncThrows`?

DougGregor replied:
Definitely a leftover, thank you!

DougGregor committed:
[Concurrency] Minor fixes for parsing and syntax of 'async'

DougGregor commented:
@swift-ci please smoke test

https://github.com/apple/swift-syntax/pull/231

DougGregor commented:
@swift-ci smoke test

apple/swift-syntax#231

DougGregor commented:
@swift-ci smoke test

apple/swift-syntax#231

DougGregor commented:
apple/swift-syntax#231

@swift-ci smoke test

DougGregor commented:
apple/swift-syntax#231

@swift-ci please smoke test

makleso6 commented:
🚀

davezarzycki commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -727,10 +727,9 @@ ERROR(generic_non_function,PointsToFirstBadToken,
-      "'throws' may only occur before '->'", ())
-ERROR(rethrows_in_wrong_position,none,
-      "'rethrows' may only occur before '->'", ())
+ERROR(async_or_throws_in_wrong_position,none,
```
Hi @DougGregor – Congrats on the first part of these change. A small nit here. It would be nice if the diagnostic selector aligned with the diagnostic name. Right now it's backwards. Specifically, the diag is named "async…" and then "…or_throws" but the first selector is "throws". The fix would be:
```
ERROR(async_or_throws_in_wrong_position,none,
      "%select{'async'|'throws'|'rethrows'}0 may only occur before '->'",
       (unsigned))
```
